staff/sergi-bittan/precourse/bingo-pro/index.js

```diff
@@ -0,0 +1,320 @@
+// StateMachine es un patrón de codigo para pasar de un estado de trabajo a otro.
+class StateMachine
+{
+    constructor()
+    {
+        this.currentState = null;
+        this.nextState = null;
+
+        var _this = this; // Capturar esta variable para poderla usar dentro de la funcion anonima
+
+        setInterval(function () 
+        {
+            if(_this.nextState != null) // Tenemos que hacer una transicion a un nuevo estado
+            {
+                if (_this.currentState != null)
+                {
+                    if(_this.currentState.OnExit != undefined)
+                    {
+                        _this.currentState.OnExit();
+                    }
+                }
+                _this.currentState = _this.nextState;
+                _this.nextState = null;
+
+                if(_this.currentState.OnEnter != undefined)
+                {
+                    _this.currentState.OnEnter();    
+                }
+                
+            }
+
+        }, 16) // run this every 60 times per second
+
+    }
+
+    GoToState(newState)
+    {
+        if(this.nextState != null)
+        {
+            console.error("La state machine esta intentando hacer varias transiciones en el mismo frame");
+        }
+
+        this.nextState = newState;        
+    }
+}
+
+//creamos una clase State para instanciar a nuestro objeto en cada estado del flujo de trabajo.
+class State
+{
+    OnEnter()
+    {
+
+    }
+    OnExit()
+    {
+
+    }
+}
+
+
+// variables globales
+var pedirNombre = null;
+var bombo = null;
+var carton = null;
+var newStateMachine = null;
+var puntos = 0;
+var ranking = [];
+
+//--------- creamos clase Carton(plantilla);
+class Carton 
+{
+    constructor() 
+    {
+        this.turnosNumero = 0;
+        this.linea = false;
+        this.numeros = [];
+        for (var i = 0; i < 15; i++) // itero para conseguir 15 
+        {
+            var numeroValido = false;
+            while (!numeroValido) 
+            {
+                var randomNumber = Math.floor(Math.random() * 99) + 1;
+                var numCarton = randomNumber;
+                if (!this.numeros.includes(numCarton)) 
+                {
+                    numeroValido = true;
+                    this.numeros.push(numCarton);
+                }
+            }
+        }
+
+    }
+    getStructureCarton() 
+    {
+        var texto = "";
+        for (var i = 0; i < this.numeros.length; i++) 
+        {
+            texto += (this.numeros[i] + " ")
+            if (i === 4 || i === 9) 
+            {
+                texto += "\n";
+            }
+        }
+        alert(texto);
+    }
+    hasLine() 
+    {
+        if (!this.linea) //ponemos la condicion para solo cantar linea una vez!
+        {
+            if (this.numeros[0] === "x" && this.numeros[1] === "x" && this.numeros[2] === "x" && this.numeros[3] === "x" && this.numeros[4] === "x") 
+            {
```

Comment:
Funciona a la perfección, aunque probablemente haya alguna manera de ahorrar poner cada uno de los casos... quizás con el método .includes(), o con un contador específico de 'x'?

Reply:
You may be right and it is something interesting to try. Let me check if I can fix during the weekend.